test: pass testSpecificMocks as function

cmd-ob opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

This PR unifies our test mocking system with extension by allowing teams to pass testSpecificMocks as a function instead of requiring the MockObject (prev TestSpecificMock). This change provides greater flexibility and control over how mocks are implemented in test files.

What Changed

Before
- Teams had to pass static mock objects to startMockServer (having to pass a `mockInstance` to `withFixtures`) or pass down this object.

After
- Teams can now pass testSpecificMocks as a function
- Functions allow dynamic mock creation and customisation
- Teams have full control over mock implementation logic

Check `e2e/MOCKING.md` for more information
## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry:

## **Related issues**

Fixes:

## **Manual testing steps**

```gherkin
Feature: my feature name

  Scenario: user [verb for user action]
    Given [describe expected initial app state]

    When user [verb for user action]
    Then [describe expected outcome]
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

cmd-ob committed:
wip

cmd-ob committed:
Refactor tests to use function based testSpecificMock

cmd-ob committed:
convert mock-server to ts and extract postRequest processing to be shared

cmd-ob committed:
refactor: update ramps API mocks and improve test-specific mock handling

cmd-ob committed:
fix linting and use update types

cmd-ob committed:
feat: enhance E2E testing with on-ramp SDK mocks and improve mock server handling

cmd-ob commented on e2e/specs/notifications/utils/mocks.ts:
```diff
@@ -99,7 +99,7 @@ interface ResponseParam {
 }
 
-async function mockAPICall(server: Mockttp, response: ResponseParam) {
+export async function mockAPICall(server: Mockttp, response: ResponseParam) {
```
Yeah i actually didnt mean to export this. But we will still keep their mock implementation as this can now be passed as a testSpecificMock function.

christopherferreira9 replied:
Since this is part of the default mocks can we clean up this file or are we still using it?

cmd-ob committed:
refactor: separate last successful build commit handling for iOS and Android

cmd-ob committed:
feat: add mock responses for on-ramp regions and update payment method handling in tests

cmd-ob committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into e2e/test-specific-mock-fn

cmd-ob committed:
feat: add favicon.ico mock response to mock server

cmd-ob committed:
feat: update mock responses to use old confirmations for various ERC20 and ERC721 tests

cmd-ob committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into e2e/test-specific-mock-fn

cmd-ob committed:
fix: update header handling in handleDirectFetch to use Headers type and improve fetch compatibility

cmd-ob committed:
add security alert send eth to quarantine

cmd-ob committed:
update withIdentityFixture to use correct pattern

cmd-ob committed:
feat: add promotional banner URL to allowlisted URLs for e2e tests

cmd-ob committed:
refactor: update type definitions in mock server and helpers for better type safety

cmd-ob committed:
Refactor API mocking in confirmation tests to use setupMockRequest

- Replaced instances of mockProxyGet with setupMockRequest in various ERC20, ERC721, and ERC1155 confirmation test files.
- Updated mock handling for Ethereum sign and personal sign tests to utilize the new setupMockRequest method.
- Enhanced error handling in account syncing tests to provide more detailed output on event emission counts.
- Adjusted mock setup in multichain account tests and quarantine tests for consistency and improved readability.
- Ensured all mock requests now specify response codes for better clarity and control over test behavior.

cmd-ob committed:
revert bitrise change

cmd-ob committed:
feat: add E2E API Mocking Guide and enforce mock server usage in tests

cmd-ob commented on shim.js:
```diff
@@ -142,5 +143,63 @@ if (enableApiCallLogs || isTest) {
+      // Patch XMLHttpRequest for Axios and other libraries
+      const OriginalXHR = global.XMLHttpRequest;
+
+      if (OriginalXHR) {
```
This allows us to track other requests through sources like Axios and other. I noticed that Detox was being help up by requests that were not showing up in the mock server.

cmd-ob commented on .github/workflows/run-e2e.yml:
```diff
@@ -25,7 +25,7 @@ jobs:
       test-suite-name: smoke-confirmations-ios
       platform: ios
-      test_suite_tag: ".*SmokeConfirmations.*"
+      test_suite_tag: ".*SmokeConfirmations(?!Redesigned).*"
```
Some ConfirmationRedesign tests were being caught in this tag

cmd-ob commented on .github/workflows/run-e2e.yml:
```diff
@@ -36,7 +36,7 @@ jobs:
       test-suite-name: smoke-confirmations-android
       platform: android
-      test_suite_tag: ".*SmokeConfirmations.*"
+      test_suite_tag: ".*SmokeConfirmations(?!Redesigned).*"
```
Some ConfirmationRedesign tests were being caught in this tag

cmd-ob committed:
refactor: comment out default security alerts mock to prevent conflicts with test-specific mocks

cmd-ob commented on e2e/api-mocking/mock-responses/cardholder-mocks.ts:
```diff
@@ -1,161 +1,110 @@
             },
           },
         },
-      ],
```
Removed a lot of fields that were not needed to reduce file sizes

christopherferreira9 replied:
The card tests are failing, are we sure the calls below (specially the eth_call) are not needed?
https://github.com/MetaMask/metamask-mobile/actions/runs/17149084987/job/48652472173?pr=18551

cmd-ob replied:
This is due to the priority thing. After adding it, the mocks from the test will "win" which is what we want. If there are any tests which need some vital mocks, it should be done in the test. At least we don't have hidden issues in the default mocks then

cmd-ob commented on e2e/pages/importSrp/ImportSrpView.ts:
```diff
@@ -28,12 +28,21 @@ class ImportSrpView {
-      elemDescription: `SRP word input at index ${srpIndex}`,
-      hideKeyboard: true,
-    });
+  async enterSrpWord(srpIndex: number, word: string): Promise<void> {
```
Android sim was adding uppercase with autocomplete

cmd-ob committed:
refactor: simplify mock server port initialization in createMockAPIServer

cmd-ob committed:
feat: add mockServerPort to createMockAPIServer return type

cmd-ob committed:
Merge branch 'main' into e2e/test-specific-mock-fn

cmd-ob committed:
feat: exclude quarantined specs from ESLint rules

cmd-ob commented on e2e/api-specs/json-rpc-coverage.js:
```diff
@@ -155,8 +156,15 @@ const main = async () => {
-    GET: [mockEvents.GET.remoteFeatureFlagsOldConfirmations],
+  const testSpecificMock = async (mockServer) => {
+    const { urlEndpoint, response } =
+      mockEvents.GET.remoteFeatureMultichainAccountsAccountDetails(false);
```
Nice catch.

cmd-ob committed:
Merge branch 'main' into e2e/test-specific-mock-fn

FrederikBolding commented on e2e/specs/snaps/test-snap-preinstalled.spec.ts:
```diff
@@ -42,7 +41,6 @@ describe(FlaskBuildTests('Preinstalled Snap Tests'), () => {
     await withFixtures(
       {
         fixture: new FixtureBuilder().withMetaMetricsOptIn().build(),
-        testSpecificMock: { POST: [mockEvents.POST.segmentTrack] },
```
Does this test still work with this removed?

cmd-ob replied:
Segment is in the default mocks now

FrederikBolding replied:
Ah, cool! Would be great to do the same for Sentry!

cmd-ob replied:
Good idea.

cmd-ob committed:
feat: enhance mock request setup with priority and update test cases

cmd-ob committed:
remove mockserver from test

cmd-ob committed:
Merge branch 'main' into e2e/test-specific-mock-fn

cmd-ob commented on e2e/specs/identity/account-syncing/multi-srp.spec.ts:
```diff
@@ -90,6 +90,9 @@ describe(SmokeIdentity('Account syncing - Mutiple SRPs'), () => {
         await Assertions.expectElementToBeVisible(WalletView.container);
+        const secretPhraseImportedText = 'Secret Recovery Phrase 2 imported';
+        await Assertions.expectTextDisplayed(secretPhraseImportedText);
+        await Assertions.expectTextNotDisplayed(secretPhraseImportedText);
```
its a toast. waiting for it to be gone before moving on to next step

cmd-ob committed:
revert confirmation pattern

mathieuartu reviewed (commented):
Insane work! Does this mean that now every auth & user storage request is mocked by default?

FrederikBolding commented on e2e/specs/snaps/test-snap-preinstalled.spec.ts:
```diff
@@ -19,6 +21,16 @@ describe(FlaskBuildTests('Preinstalled Snap Tests'), () => {
         restartDevice: true,
+        testSpecificMock: async (mockServer: Mockttp) => {
+          const { urlEndpoint, response } =
+            mockEvents.GET.remoteFeatureFlagsRedesignedConfirmationsFlask;
```
This test does not need this mock

cmd-ob committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into e2e/test-specific-mock-fn

cmd-ob committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into e2e/test-specific-mock-fn

cmd-ob commented:
> Insane work! Does this mean that now every auth & user storage request is mocked by default?

Yes it does. We will return an empty user storage for all requests and each test can mock it using `UserStorageMockttpController ™` 😺

cmd-ob committed:
test: improve toast notification handling in multi-SRP account syncing

cortisiko reviewed (dismissed):
🌮 🌮 🌮

cmd-ob commented:
E2E failures are known issues and will be fixed in a separate PR

mcmire commented on e2e/specs/snaps/test-snap-ethereum-provider.spec.ts:
```diff
@@ -17,9 +16,6 @@ describe(FlaskBuildTests('Ethereum Provider Snap Tests'), () => {
         fixture: new FixtureBuilder().withMultiSRPKeyringController().build(),
         restartDevice: true,
-        testSpecificMock: {
-          GET: [mockEvents.GET.remoteFeatureFlagsRedesignedConfirmationsFlask],
```
Is this included in the default mocks as well now?

cmd-ob replied:
@FrederikBolding  @mcmire  I could have sworn i saw another comment asking to remove the mock 😄 
I can add that back in. Do you mind allow me to get this in my next PR (which will be in today

cmd-ob replied:
Fixed

cmd-ob committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into e2e/test-specific-mock-fn

cmd-ob committed:
feat: update on-ramp SDK mocks with new light, networks, and countries responses

cmd-ob committed:
feat: add mock setup for remote feature flags in Ethereum provider snap tests

cmd-ob committed:
remove conflicting mock for tokens, remove on-ramp-sdk mocks to be moved into tests

cmd-ob committed:
refactor: add new mock to notifications defaults and remove from testspecific mock

cmd-ob committed:
feat: Add region-aware on-ramp mocks and support for Spain

- Created a new mock setup file for region-aware on-ramp API mocks.
- Updated Constants.ts to include Spain as a supported region with relevant properties.
- Modified various test specifications to utilize the new region-aware mock setup for France and Spain.
- Ensured that geolocation and API responses are tailored based on the selected region during tests.

cmd-ob committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into e2e/test-specific-mock-fn

cmd-ob committed:
fix dapp-initiated-transfer by swiping to elements, quarantine per-dapp selected
